qiskit_experiments/analysis/curve_analysis.py

```diff
@@ -0,0 +1,708 @@
+                    ),
+                ]
+
+
+        A fitting for two trigonometric curves with the same parameter
+        =============================================================
+
+        In this type of experiment, the analysis deals with two different curves.
+        However the parameters are shared with both functions.
+
+        .. code-block::
+
+            class AnalysisExample(CurveAnalysis):
+
+                __series__ = [
+                    SeriesDef(
+                        name="my_experiment1",
+                        fit_func=lambda x, p0, p1, p2, p3:
+                            cos(x, amp=p0, freq=p1, phase=p2, baseline=p3),
+                        filter_kwargs={"experiment": 1},
+                        plot_color="red",
+                        plot_symbol="^",
+                    ),
+                    SeriesDef(
+                        name="my_experiment2",
+                        fit_func=lambda x, p0, p1, p2, p3:
+                            sin(x, amp=p0, freq=p1, phase=p2, baseline=p3),
+                        filter_kwargs={"experiment": 2},
+                        plot_color="blue",
+                        plot_symbol="o",
+                    ),
+                ]
+
+
+    Notes:
+        This CurveAnalysis class provides several private methods that subclasses can override.
+
+        - Customize figure generation:
+            Override :meth:`~self._create_figures`. For example, here you can create
+            arbitrary number of new figures or upgrade the default figure appearance.
+
+        - Customize pre-data processing:
+            Override :meth:`~self._data_pre_processing`. For example, here you can
+            take a mean over y values for the same x value, or apply smoothing to y values.
+
+        - Customize post-analysis data processing:
+            Override :meth:`~self._post_processing`. For example, here you can
+            calculate new entity from fit values. Such as EPC of RB experiment.
+
+        - Customize fitting options:
+            Override :meth:`~self._setup_fitting`. For example, here you can
+            calculate initial guess from experiment data and setup fitter options.
+
+        Note that other private methods are not expected to be overridden.
+        If you forcibly override these methods, the behavior of analysis logic is not well tested
+        and we cannot guarantee it works as expected (you may suffer from bugs).
+        Instead, you can open an issue in qiskit-experiment github to upgrade this class
+        with proper unittest framework.
+
+        https://github.com/Qiskit/qiskit-experiments/issues
+    """
+
+    #: List[SeriesDef]: List of mapping representing a data series
+    __series__ = None
+
+    @classmethod
+    def _default_options(cls):
+        """Return default data processing options.
+
+        Options:
+            base_fitter: A callback function to perform fitting with formatted data.
```

Comment:
This option could probably just be called `curve_fit` or `curve_fitter`. We should document the expected/required signature of the function too.